packages/devtools_app/lib/src/screens/performance/panes/frame_analysis/frame_time_visualizer.dart

```diff
@@ -0,0 +1,325 @@
+    );
+  }
+
+  List<_FramePhaseBlockData> _generateBlockData(FrameAnalysis frameAnalysis) {
+    final frame = frameAnalysis.frame;
+    if (frame.hasShaderTime) {
+      return [
+        _FramePhaseBlockData(
+          title: 'Shader compilation',
+          duration: frame.shaderDuration,
+          flex: frameAnalysis.shaderCompilationFlex!,
+          icon: Icons.image_outlined,
+        ),
+        _FramePhaseBlockData(
+          title: 'Other raster',
+          duration: frame.rasterTime - frame.shaderDuration,
+          flex: frameAnalysis.rasterFlex!,
+          icon: Icons.grid_on,
+        ),
+      ];
+    }
+    final rasterPhase = frameAnalysis.rasterPhase;
+    return [
+      _FramePhaseBlockData(
+        title: rasterPhase.title,
+        duration: rasterPhase.duration,
+        flex: frameAnalysis.rasterFlex!,
+        icon: Icons.grid_on,
+      ),
+    ];
+  }
+}
+
+class _FrameBlockGroup extends StatelessWidget {
+  const _FrameBlockGroup({
+    Key? key,
+    required this.title,
+    required this.data,
+    required this.hasData,
+  }) : super(key: key);
+
+  final String title;
+
+  final List<_FramePhaseBlockData> data;
+
+  final bool hasData;
+
+  @override
+  Widget build(BuildContext context) {
+    Widget content;
+    if (hasData) {
+      final totalFlex =
+          data.fold<int>(0, (current, block) => current + block.flex);
+      content = LayoutBuilder(
+        builder: (context, constraints) {
+          final adjustedBlockWidths =
+              adjustedWidthsForBlocks(constraints, totalFlex);
+          return Row(
+            children: [
+              for (var i = 0; i < data.length; i++)
+                _FramePhaseBlock(
+                  blockData: data[i],
+                  width: adjustedBlockWidths[i],
+                ),
+            ],
+          );
+        },
+      );
+    } else {
+      content = const Text('Data not available.');
+    }
+
+    return Column(
+      crossAxisAlignment: CrossAxisAlignment.start,
+      children: [
+        Text(title),
+        const SizedBox(height: denseSpacing),
+        content,
+      ],
+    );
+  }
+
+  /// Returns a list of adjusted widths for each block.
+  ///
+  /// The adjusted widths will ensure each block is at least
+  /// [_FramePhaseBlock.minBlockWidth] wide, and will modify surrounding block
+  /// widths to accomodate. f
+  List<double> adjustedWidthsForBlocks(
+    BoxConstraints constraints,
+    int totalFlex,
+  ) {
+    final unadjustedBlockWidths = data
+        .map(
+          (blockData) => constraints.maxWidth * blockData.flex / totalFlex,
+        )
+        .toList();
+
+    var adjustment = 0.0;
+    var widestBlockIndex = 0;
+    for (var i = 0; i < unadjustedBlockWidths.length; i++) {
+      final unadjustedWidth = unadjustedBlockWidths[i];
+      final currentWidestBlock = unadjustedBlockWidths[widestBlockIndex];
+      if (unadjustedWidth > currentWidestBlock) {
+        widestBlockIndex = i;
```

Comment:
if all blocks are less than the min width, then our widest block will end up being less than the min width. Is that ok based on the context with these block resizes?

Reply:
Below where we calculate the adjustedWidth, we adjust so that the width is at least `_FramePhaseBlock.minBlockWidth`:

`adjustedWidth = math.max(adjustedWidth, _FramePhaseBlock.minBlockWidth);`

So it won't end up being less than the min width after the adjustment. And if we overflow when all blocks are ` _FramePhaseBlock.minBlockWidth` wide, then so be it, because that means the screen is more narrow than we intend to support.